src/snowflake/cli/api/project/schemas/updatable_model.py

```diff
@@ -14,17 +14,147 @@
+
+
+def _is_templated(info: ValidationInfo, value: Any) -> bool:
+    return (
+        info.context
+        and info.context.get("skip_validation_on_templates", False)
+        and isinstance(value, str)
+        and PROJECT_TEMPLATE_START in value
+    )
+
+
+_initial_context: ContextVar[Optional[Dict[str, Any]]] = ContextVar(
+    "_init_context_var", default=None
+)
+
+
+@contextmanager
+def context(value: Dict[str, Any]) -> Iterator[None]:
+    """
+    Thread safe context for Pydantic.
+    By using `with context()`, you ensure context changes apply
+    to the with block only
+    """
+    token = _initial_context.set(value)
+    try:
+        yield
+    finally:
+        _initial_context.reset(token)
+
 
 class UpdatableModel(BaseModel):
     model_config = ConfigDict(validate_assignment=True, extra="forbid")
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(**kwargs)
+    def __init__(self, /, **data: Any) -> None:
+        """
+        Pydantic provides 2 options to pass in context:
+        1) Through `model_validate()` as a second argument.
+        2) Through a custom init method and the use of ContextVar
+
+        We decided not to use 1) because it silently stops working
+        if someone adds a pass through __init__ to any of the Pydantic models.
+
+        We decided to go with 2) as the safer approach.
+        Calling validate_python() in the __init__ is how we can pass context
+        on initialization according to Pydantic's documentation:
+        https://docs.pydantic.dev/latest/concepts/validators/#using-validation-context-with-basemodel-initialization
+        """
+        self.__pydantic_validator__.validate_python(
+            data,
+            self_instance=self,
+            context=_initial_context.get(),
+        )
+
+    @classmethod
+    def _is_entity_type_field(cls, field: Any) -> bool:
+        """
+        Checks if a field is of type `EntityTypeField`
+        """
+        if not isinstance(field, FieldInfo) or not field.json_schema_extra:
+            return False
+
+        return (
+            "is_type_field" in field.json_schema_extra
+            and field.json_schema_extra["is_type_field"]
+        )
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        """
+        This method will collect all the Pydantic annotations for the class
+        currently being initialized (any subclass of `UpdatableModel`).
+
+        It will add a field validator wrapper for every Pydantic field
+        in order to skip validation when templates are found.
+
+        It will apply this to all Pydantic fields, except for fields
+        marked as `EntityTypeField`. These will be skipped because
+        Pydantic does not support validators for discriminator field types.
+        """
+
+        super().__init_subclass__(**kwargs)
+
+        field_annotations = {}
+        field_values = {}
+        # Go through the inheritance classes in reverse order
```

Comment:
nit: explain why reverse order here please (I remember from our conversation because it was only last week, but I don't trust I'll immediately remember later).